### PR TITLE
asyncnet: don't try to close the socket again [backport]

### LINKED ISSUE #9867: Long-lived application occasionally dies with SIG_PIPE when calling httpclient.request
Our serverside application, currently on 0.18, does http the occasional http request to the Google APIs. Sometimes, I think, google closes the connection on us and the application dies with SIG_PIPE:

```
...
keyprovision_google.nim(66) getAccessToken
keyprovision_google.nim(30) requestJson
keyprovision_google.nim(71) :anonymous
httpclient.nim(1196)     request
httpclient.nim(1181)     requestAux
httpclient.nim(1031)     parseResponse
httpclient.nim(883)      close
net.nim(866)             close
SIGPIPE: Pipe closed.
```

Of course, the http request() is wrapped into try/except; that does not catch the signal however.

This is running on nim 0.18 (I can't try 0.19 yet, porting it is significant effort due to libraries used), and in a docker container on Linux.

I found this thread: https://forum.nim-lang.org/t/4417, but that didn't really help. The httpclient is already configured to use a reasonable timeout (3s).

I'm fairly certain this is because Google is closing the kept-alive connection in the background; but I'm hesitant to add `Connection: Close` (or otherwise close the connection on our end) unless that's the only way to solve this.

### PR DESCRIPTION
**Pull request by alaviss**
The closed flag here doesn't seem to be a a good design to me, but until
I can understand why it was needed it seems unwise to just replace it
with the invalidation of the socket.

Fixes https://github.com/nim-lang/Nim/issues/9867#issuecomment-671071188

/cc @dom96 as this `closed` flag was added by him. This flag appear to have survived the refactoring in https://github.com/nim-lang/Nim/commit/1661062ebf147e55e29a0bfb8fc62dc7959b8614, where its sister `net` module use the different approach of invalidating the `Socket` instead of a flag.

### CODE REVIEW

File: lib/pure/asyncnet.nim [hunk 1]
```diff
@@ -713,25 +713,27 @@ elif defined(nimdoc):
...
-          SSL_shutdown(socket.sslHandle)
-        else:
-          0
-      SSL_free(socket.sslHandle)
-      if res == 0:
-        discard
-      elif res != 1:
-        raiseSSLError()
-  socket.closed = true # TODO: Add extra debugging checks for this.
+  if not socket.closed:
```
Comment (dom96) [inline lib/pure/asyncnet.nim]: Prefer: 
```
if socket.closed: return
```

Style rather than wrapping everything in more and more nested ifs.

**Review by dom96 — CHANGES_REQUESTED** [review]
IMO an exception should be raised if you try to close an already closed socket. Maybe you can raise an exception instead of doing nothing?

**Comment by alaviss** [comment]
> IMO an exception should be raised if you try to close an already closed socket. Maybe you can raise an exception instead of doing nothing.

`net` currently doesn't raise any exception so we shouldn't raise any here either.

**Comment by dom96** [comment]
it always calls `close` on the fd though, maybe that will raise?

**Comment by alaviss** [comment]
Nah, `nativesockets.close()` ignores all errors.

**Comment by dom96** [comment]
That sounds dangerous, but alright, in any case I would still match the semantics in `net` and call `close` nevertheless.

**Comment by alaviss** [comment]
Currently I'm not doing that because I'm not invalidating the socket. I can certainly do so, but IMO it's better to do that with the follow up refactoring to get rid of the `closed` flag.

File: lib/pure/asyncnet.nim (hunk 1 above)
Comment (Araq) [reply lib/pure/asyncnet.nim]: No, nested ifs are better.

**Pull request merged**